[RFC]: Introduce DMA domain on Zephyr

LaurentiuM1234 opened the pull request:
This is a very rough attempt to implement the DMA domain on Zephyr. Most of it is inspired from dma_multi_chan_domain.

Initially, I tried using the same approach as zephyr_domain.c as in having a Zephyr thread run for each core and said threads would do their assigned work if the DMA IRQs would give them semaphore resources but this approach has the following limitations:

1) Because of how task canceling is done ( `zephyr_ll_task_cancel()` sets the task's state to CANCELED and `zephyr_ll_run()` makes sure to call `domain_unregister()` if task as canceled) this leads to the same problem described in second to last patch (
`schedule: zephyr_ll: Call zephyr_ll_task_done on zephyr_ll_task_cancel` ).
2) Using aforementioned patch with the thread approach would probably not be very good as that would probably mean (I am not 100% sure) calling `k_thread_abort()` from IRQ context.

In order to avoid the second limitation (first one still remains, sadly) I chose an approach similar to dma_multi_chan_domain as in zephyr_ll_run will be executed from DMA IRQ context.

I noticed that `domain_is_pending()` was entirely removed from zephyr_ll.c. Without it, I believe DMA domain will not be able to work properly (because, in the case of DMA domain, not all tasks should be executed whenever we receive a DMA IRQ so `domain_is_pending()` helps out with that). If you think this solution is viable I think the next step would be to bring back `domain_is_pending()`.

There's also the fact that zephyr_domain's `unregister()` function will call `k_thread_abort()` from IRQ context (because of this patch series) which is the same problem as DMA domain approach with Zephyr threads. I am curious as to why was the threads approach chosen over the IRQ approach for timer domain.

Currently, I've tested this solution using imx-stable-v2.2 branch and it seems to work fine. (tested doing consecutive aplays) Will also have to test using main branch.

For now, only the case where we have 1 PPL task at a time will work.

lyakh commented on src/schedule/zephyr_dma_domain.c:
```diff
@@ -0,0 +1,298 @@
+#define interrupt_disable mux_interrupt_disable
+#endif
+
+struct driving_channel {
```
maybe it would be good to keep it a bit similar to what we have now and call this `struct dma_domain_data` or if you want easy grepping `struct dma_zephyr_domain_data`

lyakh commented on src/schedule/zephyr_dma_domain.c:
```diff
@@ -0,0 +1,298 @@
+	/* pipeline task which registered the DMA IRQ */
+	struct pipeline_task *ptask;
+	bool in_use; /* true if channel is currently in use */
+	void (*work)(void *work_data); /* work to be executed by DMA IRQ */
```
this could still be `handler`

lgirdwood replied:
can we use the same terminology as zephyr LL.

LaurentiuM1234 replied:
sure, will do

lyakh commented on src/schedule/zephyr_dma_domain.c:
```diff
@@ -0,0 +1,298 @@
+	struct pipeline_task *ptask;
+	bool in_use; /* true if channel is currently in use */
+	void (*work)(void *work_data); /* work to be executed by DMA IRQ */
+	void *work_data; /* data to be used by work function */
```
and this `arg`

lyakh commented on src/schedule/zephyr_dma_domain.c:
```diff
@@ -0,0 +1,298 @@
+	struct driving_channel *driving = data;
+
+	/* do work */
+	driving->work(driving->work_data);
```
I don't think it's a good idea to return to executing the audio (LL) scheduler in interrupt context again. I read your explanation in the description to the PR but I'd still rather keep the scheduler as a Zephyr thread. We want it to be interruptible. I don't know what design considerations you have on iMX, but wouldn't having interruptible per-core scheduling threads be beneficial for you too?

lgirdwood replied:
Ack - we cant run this in IRQ context. Zephyr is a general purpose RTOS and uses thread priority levels to determine priority for scheduling (xtos used IRQ levels to determine priority).

lgirdwood commented on src/schedule/zephyr_dma_domain.c:
```diff
@@ -0,0 +1,298 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
I would call this zephyr_preempt_domain.c because the code should be reusable for timers too to schedule processing that is not low latency. i.e. media decode with large buffers.

lyakh replied:
@lgirdwood sorry, could you explain? As far as I understand the SOF LL scheduler has a "domain" abstraction which is a mechanism that triggers the LL scheduler scheduling function upon an event. One of those domains is a timer domain. This PR introduces a new *alternative* domain - a DMA domain. How can "timers" also use it? Isn't it decided with a task granularity - you initialise a task with a specific domain and then that domain handles the task?

lgirdwood replied:
@lyakh I mean the timer should also be able to schedule work in the preemptive domain (alongside the LL domain).

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -444,7 +426,7 @@ static int zephyr_ll_task_cancel(void *data, struct task *task)
 	zephyr_ll_lock(sch, &flags);
 	if (task->state != SOF_TASK_STATE_FREE)
-		task->state = SOF_TASK_STATE_CANCEL;
+		zephyr_ll_task_done(sch, task);
```
Btw, I think we probably have a chance after we align on this PR to simplify some of the scheduler APIs for Zephyr usage (since we will have a good understanding of what we need to support timers, dma as scheduling triggers).

LaurentiuM1234 replied:
@lyakh `domain_register()`/`domain_unregister()` is called for every task, otherwise DMA domain wouldn't even have a chance of working since `domain_register()` is the one enabling the DMA IRQs (I can provide screenshots of the logs showing `domain_unregister()`/`domain_register()` being called multiple times if need be). Since we can have multiple DMA IRQs `domain_register()` should and is called for each task. What makes you say it's only called for first/last tasks? Also, that's the only logical explanation I could find for the following behavior:

1) Initially, when a playback and a capture are scheduled the capture calls  `domain_unregister()` when finished and works fine but playback doesn't call `domain_unregister()` so, scheduling another playback immediately after this doesn't work.
2) Trying to schedule a playback immediately after another playback doesn't work since the first playback doesn't call `domain_unregister()`.
3) Trying to schedule a capture after a playback works fine since they use different channels so `zephyr_ll_run` is still able to be executed thanks to the fact that capture's DMA IRQs are still enabled.

LaurentiuM1234 commented on zephyr/CMakeLists.txt:
```diff
@@ -550,6 +550,7 @@ zephyr_library_sources(
 	${SOF_SRC_PATH}/init/ext_manifest.c
 	${SOF_SRC_PATH}/arch/xtensa/drivers/cache_attr.c
 	${SOF_SRC_PATH}/schedule/zephyr_domain.c
+	${SOF_SRC_PATH}/schedule/zephyr_dma_domain.c
```
ack

LaurentiuM1234 commented on src/include/sof/schedule/ll_schedule.h:
```diff
@@ -34,7 +34,7 @@ struct ll_task_pdata {
 };
 
-#if !defined(__ZEPHYR__) || defined(CONFIG_IMX)
+#if !defined(__ZEPHYR__)
```
ack

LaurentiuM1234 commented on src/include/sof/schedule/ll_schedule_domain.h:
```diff
@@ -191,17 +191,20 @@ static inline bool domain_is_pending(struct ll_schedule_domain *domain,
+#define timer_domain_init(timer, clk) zephyr_domain_init(clk)
+#define dma_multi_chan_domain_init(dmas, num_dmas, clk, aggregated_irq)\
+	zephyr_dma_domain_init(dmas, num_dmas, clk)
+#endif
```
ack

LaurentiuM1234 commented on src/platform/imx8/platform.c:
```diff
@@ -212,7 +211,6 @@ int platform_init(struct sof *sof)
 
 	/* show heap status */
 	heap_trace_all(1);
-#endif /* __ZEPHYR__ */
```
ack

LaurentiuM1234 commented:
> Thanks @LaurentiuM1234 good to see this making progress :)
> 
> > 1. Because of how task canceling is done ( `zephyr_ll_task_cancel()` sets the task's state to CANCELED and `zephyr_ll_run()` makes sure to call `domain_unregister()` if task as canceled) this leads to the same problem described in second to last patch (
> 
> IIRC, for LL tasks we set the cancel state and this is then checked at next LL thread run. I would imagine we could do the same here, i.e.
> 
> 1. thread is processing
> 2. IPC comes in to STOP
> 3. IPC sets pipeline state
> 4. thread processing complete, logic then check state and cleans up pipeline and stops thread.
> 
> > `schedule: zephyr_ll: Call zephyr_ll_task_done on zephyr_ll_task_cancel` ).
> > 2. Using aforementioned patch with the thread approach would probably not be very good as that would probably mean (I am not 100% sure) calling `k_thread_abort()` from IRQ context.
> > In order to avoid the second limitation (first one still remains, sadly) I chose an approach similar to dma_multi_chan_domain as in zephyr_ll_run will be executed from DMA IRQ context.
> 
> We should be able to use the same mechanism that we use to start the LL thread for timer based triggering.

@lgirdwood The problem is that when the IPC comes with the TRIGGER STOP the SAI stops and so does the DMA IRQ so there's no more IRQs to give resources to the semaphore in the Zephyr thread to actually do its work. Because of this, the DMA IRQ (from EDMA's POV) never gets unregistered (since the unregistering process is done in the thread's work function). I will try to see if I can find a way to overcome this but if not I think this would be an ok approach for now. Also, the same should have applied to timer domain if it weren't for the DMA trace task being scheduled permanently.

Also, I believe we should reintroduce `domain_is_pending()`. As of now, whenever we get a DMA IRQ all the tasks are executed which is wrong. This will allow us to only execute the tasks which are associated with the DMA IRQ that came. Let me know what you think.

lgirdwood commented:
> @lgirdwood The problem is that when the IPC comes with the TRIGGER STOP the SAI stops and so does the DMA IRQ so there's no more IRQs to give resources to the semaphore in the Zephyr thread to actually do its work. Because of this, the DMA IRQ (from EDMA's POV) never gets unregistered (since the unregistering process is done in the thread's `work` function). I will try to see if I can find a way to overcome this.

That's a good point - we have to assume that the last DMA IRQ can occur after the the DMA is stopped/drained and always after the STOP IPC. We probably need some logic to check in the task status in the domain IRQ handler.

LaurentiuM1234 commented:
V2 changes:

1) Switched from the IRQ-based approach to the thread-based approach.
2) Removed the commits that were integrating the DMA domain. I believe that, for now, the main focus of this PR should be on the commits that implement the DMA domain. Once the solution becomes more viable, I will reintroduce the other commits, also addressing @lyakh 's suggestions.

**Problems**

1. When trying to do 2 consecutive `aplay` the second one doesn't work properly, resulting in an input/output error. I did some digging around and it seems like after a `sai_stop()`/ operation I need to do `interrupt_disable()` and `interrupt_unregister()` before the next `sai_start()` operation. As of now, I couldn't find an explanation for this behavior.
2. In the aforementioned case, if I do a `zephyr_ll_task_done()` operation inside a `zephyr_ll_task_cancel()` operation it seems like that fixes the issue. The problem is, the DMA trace acts somewhat weird. When the trigger STOP operation comes, the DMA trace should normally print some logs (e.g: `edma_stop`, `sai_stop`) immediately. It seems like, these logs get printed only when the second aplay comes around.

lyakh reviewed (commented):
looks ok, thanks for updating

lgirdwood reviewed (commented):
@LaurentiuM1234 might be worth adding a user to an Intel nocodec topology2 as a subsequent PR. This way it could be tested within the Intel CI as we have no way of testing in CT atm ?

LaurentiuM1234 commented:
> @LaurentiuM1234 might be worth adding a user to an Intel nocodec topology2 as a subsequent PR. This way it could be tested within the Intel CI as we have no way of testing in CT atm ?

Will try to have a look a bit later. As of now, it seems like the solution is not very stable anyways.

dbaluta commented:
> It sounds like this would be good as an staging or experimental feature once you are happier with maturity? That way we could get more developers to test and eliminate the issues. Btw, what issues are you seeing ? @kv2019i fyi

That's a very hard one. After stress-testing ~ 30 mins, with aplay/arecord the firmware looks like to be crashing.

lgirdwood commented:
> > It sounds like this would be good as an staging or experimental feature once you are happier with maturity? That way we could get more developers to test and eliminate the issues. Btw, what issues are you seeing ? @kv2019i fyi
> 
> That's a very hard one. After stress-testing ~ 30 mins, with aplay/arecord the firmware looks like to be crashing.

So we are all productizing SOF on Zephyr on xtensa and this I think would be a good stress test (under a experimental/staging) Kconfig.

This way it can be shared with more SOF and Zephyr developers as a scheduling test case and also be stressed by CI. This will help us identify any lingering core logic issues.

I'm keen to merge, providing the Kconfig has the flag and the warning. I'd like to help and see this working well.

dbaluta commented on src/include/sof/schedule/ll_schedule.h:
```diff
@@ -34,7 +34,7 @@ struct ll_task_pdata {
-#if !defined(__ZEPHYR__) || defined(CONFIG_IMX)
+#if !defined(__ZEPHYR__) || (defined(CONFIG_IMX) && !CONFIG_PREEMPT_DOMAIN)
 int scheduler_init_ll(struct ll_schedule_domain *domain);
 
```
why not use `!defined(CONFIG_PREEMPT_DOMAIN)` ?

LaurentiuM1234 replied:
because CONFIG_PREEMPT_DOMAIN is technically defined when on Zephyr (please see zephyr/Kconfig) so we need to actually check if it was set to y or no

LaurentiuM1234 commented:
**V3 changes**:

1) Previously, because of the fact that DMA IRQs would get cut off whenever a TRIGGER STOP would come the preempt domain thread would stop doing its assigned work. Because of the fact that zephyr_ll's `zephyr_ll_task_cancel()` would not remove the task from the task queue whenever it was called that lead to DMA IRQs remaining masked (because of the fact that SAI's `set_config()` function would mask them every time it was called). In order to overcome this, I introduced a new domain function `domain_task_cancel()`. The purpose of this function is to let the preempt domain know the number of remaining tasks when `zephyr_ll_task_cancel()` is called. Whenever that count reaches 0, the domain will know that the DMA IRQs got cut off. In order to avoid aforementioned problem, `domain_task_cancel()` gives a resource to the thread semaphore in order to allow it to execute its assigned work one more time after the DMA IRQs get cut off. This will allow the thread to remove the task from the task queue and call `domain_unregister()`.

2) Solved

> In the aforementioned case, if I do a zephyr_ll_task_done() operation inside a zephyr_ll_task_cancel() operation it seems like that fixes the issue. The problem is, the DMA trace acts somewhat weird. When the trigger STOP operation comes, the DMA trace should normally print some logs (e.g: edma_stop, sai_stop) immediately. It seems like, these logs get printed only when the second aplay comes around.

by simply switching to the main branch to do the testing. It turns out that the problem was caused by the fact that the queued IPCs were sent whenever `zephyr_ll_task_run()` was executed. Due to the fact that the DMA IRQs got cut off, the queued IPCs would not be sent until the DMA IRQs would get enabled again.

3) Preempt domain is now added as an experimental feature. This means that the developers can simply turn on CONFIG_PREEMPT_DOMAIN in order to test it out. Until this becomes more stable, I believe this might be the way to go.

**Problems**:

1) Currently there seems to be something weird going on whenever I enable DMA tracing with preempt domain. This will require further investigations. For now, it seems that if the DMA tracing is disabled the current version passes the stress testing (which is just a series of aplay's and arecords's having a 3s sleep in between them). The problem consists of the fact that the kernel doesn't get a response for some IPCs so it gets a timeout.

**TODOs**:

1) Add `domain_is_pending()` to zephyr_ll and come up with a decent implementation of it inside preempt domain. (The current one will probably not work for mixer topologies for example)
2) Figure out why Problem 1 occurs.
3) Make sure that all Zephyr threads started by preempt domain are aborted at some point.

LaurentiuM1234 commented:
> @LaurentiuM1234 this looks good to me. We can merge it so that other also get a chance to test it.

sounds good to me.

sorry, closed PR by mistake.

kv2019i commented on src/schedule/zephyr_preempt_domain.c:
```diff
@@ -0,0 +1,500 @@
+		return ret;
+	}
+
+	/* TODO: thread abortion logic goes here */
```
This is probably needed for merging, or...? The logic around dt->handler seems a bit suspicious. When unregistering a domain, handler is not clear, but yet in register we use this field to decide whether to start a new thread. Does this handle multiple register/unregister cycles in its current form? I don't think we need to have everything polished to merge this, but just looking for a sane minimal-functionality set...

LaurentiuM1234 replied:
I agree this look a bit shady.

When I did the testing it seemed to work fine without it. The tests consisted of a bunch of `aplay`/`arecord` having a 3s sleep in between them in order to make the FW reboot + some `aplay` while doing a lot of pause pushing. Also, the DMA trace was disabled during the tests. Also, I did a few consecutive `aplay` (without the FW boot) and it seemed to work fine as well.

I can't guarantee that this will not cause problems on systems with multiple cores as I haven't been able to test that scenario.

This probably works after FW reboot (which should be the only problematic case) because of the fact that `dt->handler` is made NULL by preempt domain's initialization function (which from my understanding should be called after every FW boot)

kv2019i commented on src/include/sof/schedule/ll_schedule_domain.h:
```diff
@@ -41,6 +41,7 @@ struct ll_schedule_domain_ops {
 	void (*domain_clear)(struct ll_schedule_domain *domain);
 	bool (*domain_is_pending)(struct ll_schedule_domain *domain,
 				  struct task *task, struct comp_dev **comp);
+	void (*domain_task_cancel)(struct ll_schedule_domain *domain, uint32_t num_tasks);
```
This should be early in the seriess, right? I mean I think the previous patch already used this, so this would break bisect.

LaurentiuM1234 replied:
You're right, this should definitely be added before the introduction of the preempt domain, thanks!

kv2019i reviewed (commented):
Looks pretty good @LaurentiuM1234 ! A few minor notes, but it would seems this would soon be in mergeable state.

LaurentiuM1234 commented:
**V4 changes**:

1) Solved problem with the order of the commits pointed out by @kv2019i. Now the introduction of `domain_task_cancel()` comes before its use in preempt domain.
2) Removed `domain_is_pending()` and all its logic since it's not being used right now. Will reintroduce this later on.

lgirdwood commented:
> **V4 changes**:
> 
> 1. Solved problem with the order of the commits pointed out by @kv2019i. Now the introduction of `domain_task_cancel()` comes before its use in preempt domain.
> 2. Removed `domain_is_pending()` and all its logic since it's not being used right now. Will reintroduce this later on.

@LaurentiuM1234 did you fix the stability, fwiw a critical Zephyr (xtensa?) schedule/timer fix was merged recently that may help. @kv2019i will @LaurentiuM1234 need the fix ?

dbaluta commented:
> @LaurentiuM1234 did you fix the stability, fwiw a critical Zephyr (xtensa?) schedule/timer fix was merged recently that may help. @kv2019i will @LaurentiuM1234 need the fix ?

It is very strange what is happening. with the stability test.

Without logging activated (e.g tracing work scheduled) everything works fine for 12 hrs. 

With logging activated at some point it looks like an IPC gets a reply, but the reply header is wrong. 

@lgirdwood @kv2019i care to point us to the fix?

LaurentiuM1234 commented:
@lgirdwood @dbaluta Just did a test with the latest changes brought to main branch and the IPC problem is still there

lgirdwood commented:
@dbaluta @LaurentiuM1234 we have a similar IPC corruption issue in IPC4 mode. We suspect the host or another module may be overwriting the mailbox during processing (i.e. logging, trace or stream status). i.e. we may need to either lock mailbox IO or copy the mailbox data during IPC processing and copy back for IPC reply. @lyakh fyi

lgirdwood commented:
> With logging activated at some point it looks like an IPC gets a reply, but the reply header is wrong.

@dbaluta @LaurentiuM1234 Btw, what header do you get and what do you expect ? Could help in solving. How long is playback before the problem occurs.

One more thing, what happens if verbose logging is enabled ? does the reproduction rate change. i.e. can we make it easy to reproduce. If easy to repor, it may be worthwhile to change an Intel topology in this PR to test the code and see what teh result is on Intel (as this could also shed more light on the root cause).

LaurentiuM1234 commented:
> Btw, what header do you get and what do you expect ? Could help in solving. How long is playback before the problem occurs.

It seems as if the host receives the exact amount of bytes it's trying to send to the DSP. Here's a sample of the error message:

```
reply size (100) exceeds the buffer size (20)
ipc tx error for 0x30010000 (msg/reply size: 100/20): -22
Failed to setup widget PGA2.0
Failed to restore pipeline after resume -22
ASoC: error at snd_soc_pcm_component_pm_runtime_get
```

> One more thing, what happens if verbose logging is enabled ? does the reproduction rate change. i.e. can we make it easy to reproduce.

Just did a test and it doesn't seem to change anything in particular. The problem seems to occur just as fast as in the case without verbose tracing. The max time I was able to run the test without the problem occurring was around 20 mins if I remember correctly? Usually it happens really fast (e.g: it can even happen in the beginning when we schedule the capture and playback tasks together)

It seems to me that the problem pointed out by @dbaluta is not the only one. I've also had some runs in which the IPCs would get a timeout (even with 10.000 ms timeout in the kernel) and when I inspected the logs it seemed to me as if they were frozen. The steps would be the following:

1) Start the sof-logger in the beginning, while the playback and capture tasks are still running.
2) Get a kernel error saying some IPC timed out.
3) Get the following error from sof-logger and crash:
`error: in logger_read(), fread(..., /sys/kernel/debug/sof/trace) failed: Operation not permitted(-1)`
4) Restart sof-logger => the logger would show the logs printed by the timer Zephyr thread from `zephyr_domain`

The problem seems to occur only after a FW reboot so if I were to run a bunch of `aplay`s without allowing the FW to suspend (because of it being inactive for 2s) the problem wouldn't occur at all.

Also, usually, the problem only happens in the beginning or in the end whenever I do an `aplay` after FW gets suspended (i.e: around the time the buffers get connected/disconnected but it can occur even sooner than the buffer connection IPCs)

If it helps, the test script I used has the following pseudocode:

```
for i = 1, 1000 do
    for j = 1, 10 do
        aplay -d 2 sample.wav
        sleep 3s
    done
done
```

> If easy to repor, it may be worthwhile to change an Intel topology in this PR to test the code and see what teh result is on Intel (as this could also shed more light on the root cause).

If the aforementioned problems don't occur on Intel's side then I'm somewhat reluctant to add a new topology as it probably wouldn't be of much help.

I think the IPC problem might not be related to the introduction of the preempt domain (I'm not sure if this is relevant but I also did a run in which I commented out the line from `platform.c` which added the `zephyr_ll` scheduler with type `SOF_SCHEDULE_LL_DMA` to the scheduler list thus not making use of the preempt domain and the aforementioned problems still occurred). If it's alright with you @dbaluta @lgirdwood @lyakh @kv2019i I believe we can merge this PR for now as it seems to work okay without enabling the DMA trace.

lgirdwood commented:
> > Btw, what header do you get and what do you expect ? Could help in solving. How long is playback before the problem occurs.
> 
> It seems as if the host receives the exact amount of bytes it's trying to send to the DSP. Here's a sample of the error message:
> 
> ```
> reply size (100) exceeds the buffer size (20)
> ipc tx error for 0x30010000 (msg/reply size: 100/20): -22
> Failed to setup widget PGA2.0
> Failed to restore pipeline after resume -22
> ASoC: error at snd_soc_pcm_component_pm_runtime_get
> ```

oh, what is the use case. I'm confused by why this error would occur during playback ?

Agree - can be merged.

lgirdwood commented on src/schedule/zephyr_preempt_domain.c:
```diff
@@ -0,0 +1,460 @@
+	struct zephyr_preempt_domain_data *data;
+	k_tid_t thread;
+	int core, ret;
+	char thread_name[] = "ll_thread0";
```
Should this be const, i.e. should we base this on the pipeline name ?

LaurentiuM1234 replied:
I'm not sure that would work for now as we can have multiple pipeline tasks running on the same Zephyr thread so I think a generic name is more suitable.

lgirdwood commented on src/schedule/zephyr_preempt_domain.c:
```diff
@@ -0,0 +1,460 @@
+		k_sem_give(sem);
+
+	/* clear IRQ */
+	dma_interrupt_legacy(channel, DMA_IRQ_CLEAR);
+	interrupt_clear_mask(irq, BIT(channel_index));
```
Should we not do these two steps before the k_sem_give() so that interrupts can continue to be sent. e.g. if the thread gets stuck we can at least get the next IRQ, check the thread status and then print an error and/or take other action.

LaurentiuM1234 replied:
Makes sense, will change.

lgirdwood commented on src/schedule/zephyr_preempt_domain.c:
```diff
@@ -0,0 +1,460 @@
+	interrupt_clear_mask(irq, BIT(channel_index));
+}
+
+static int enable_dma_irq(struct zephyr_preempt_domain_data *data)
```
this could be timer, can we rename ```_dma_``` to ```_prempt_```

lgirdwood replied:
unless this code is dma specific ? and we need a similar function to pass in another wake source

LaurentiuM1234 replied:
The only similarity between let's say a timer enable function and the DMA one is the fact that we need to register a handler and then enable the IRQ. For now, I say we leave this as is so as to avoid any confusions and we can change it later on when the support for timer IRQs will be added.

LaurentiuM1234 commented:
> oh, what is the use case. I'm confused by why this error would occur during playback ?

It probably happened in the beginning when the playback and capture pipeline tasks are scheduled or during an `arecord` done after a FW resume.

I'm very sorry, for some reason I only included a portion of the pseudocode of the test script. The script (pseudocode) actually looks like this: 

```
for i = 1, 1000 do
    for j = 1, 10 do
        aplay -d 2 sample_playback.wav
        sleep 3s
    done

    for j = 1, 10 do
        aplay -d 2 sample_playback.wav
        arecord -d 2 sample_capture.wav
        sleep 3s
    done

    for j = 1, 10 do
        arecord -d 2 sample_capture.wav
        sleep 3s
    done
done
```

Will also edit the initial comment to include the full pseudocode.

lgirdwood commented on src/schedule/zephyr_preempt_domain.c:
```diff
@@ -0,0 +1,460 @@
+				 dt,
+				 NULL,
+				 NULL,
+				 CONFIG_NUM_PREEMPT_PRIORITIES - 1,
```
I think we can read priority from topology. i.e. say we have 2 DMA or timer based preemptable threads we may want one to have priority over the other.

LaurentiuM1234 replied:
I agree with you but for now I think it's best to leave the priority as-is since it's not really a priority to have the threads priorities based on the pipelines. Also, this would beg the following question: what priority should the thread have if we try to schedule multiple pipeline tasks having different priorities using the same thread?

dbaluta commented:
> oh, what is the use case. I'm confused by why this error would occur during playback ?

When playback starts (after SOF device entered pm runtime suspend, ALSA needs to pm runtime resume the device. 

So, 

```
sof_resume
 ->    ret = tplg_ops->set_up_all_pipelines(sdev, false);
       -> sof_ipc3_set_up_all_pipelines
           -> sof_ipc3_widget_setup
                ->  hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW; 
                 -> hdr.size = sizeof(volume->config); // this size is 100
               -> sof_ipc_tx_message(sdev->ipc, swidget->private, hdr->size, &reply, sizeof(reply))
```

Now, when te reply arrives one would expect reply->hdr.size to be sizeof(reply). But for some reason, the message read from input mailbox looks to be *not* a reply but actually the original message sent.

So, the DSP sends an interrupt saying that it has put a reply mesage for the Host but actually doesn't modify the message. 

It looks very strange and I'm trying to understand if there is some sort of caching issue (?) but it doesn't look so, or the DSP actually triggers an unwanted interrupt which the host identifies as a reply.

lgirdwood commented on src/schedule/zephyr_preempt_domain.c:
```diff
@@ -0,0 +1,460 @@
+#endif
+
+#define SEM_LIMIT 1
+#define ZEPHYR_LL_STACK_SIZE 8192
```
We need to change the naming to remove LL to avoid confusion.

lgirdwood reviewed (commented):
Just some minor questions - I think next steps (in a subsequent PR) would be to support 
1) Pass stack size in with the topology
2) N threads per core (using Zephyr dynamic stack allocator).

lyakh commented:
> If it's alright with you @dbaluta @lgirdwood @lyakh @kv2019i I believe we can merge this PR for now as it seems to work okay without enabling the DMA trace.

@LaurentiuM1234 @lgirdwood sorry I still don't understand why this is now called "preempt domain," isn't it doing the same as what the DMA domain is doing with XTOS?

LaurentiuM1234 commented:
> @LaurentiuM1234 @lgirdwood sorry I still don't understand why this is now called "preempt domain," isn't it doing the same as what the DMA domain is doing with XTOS?

You are right. For now, `preempt domain` works just as the DMA domain in XTOS.

My current understanding is that, thanks to the fact that `preempt domain` is pretty generic (as in it it registers a DMA interrupt (which can be any interrupt if we generalize the function that does the interrupt - here, `register_dma_irq`) creates a thread and then based on that interrupt the thread does its assigned work) we can also use it with multiple interrupts, not just the ones from the DMA.

As @lgirdwood pointed out, we can also use it with timers (although I for one am not sure what's the difference between using `zephyr_domain` and `zephyr_preempt_domain` with a timer)

dbaluta commented:
> . If easy to repor, it may be worthwhile to change an Intel topology in this PR to test the code and see what teh result is on Intel (as this could also shed more light on the root cause).

@lgirdwood what Intel topology would be the most fitted for us to try to enable preempt domain on? There are dozens of them :) and we don't know which one is easier for you to use.

LaurentiuM1234 commented:
**V5 changes**

1) Changed name of the Zephyr thread to "pdomain_thread". This should give it a more generic name and remove the LL naming.
2) Changed ZEPHYR_LL_STACK_SIZE to ZEPHYR_PDOMAIN_STACK_SIZE.
3) Changed order of operations inside the DMA IRQ handler. Reason:

> Should we not do these two steps before the k_sem_give() so that interrupts can continue to be sent. e.g. if the thread gets stuck we can at least get the next IRQ, check the thread status and then print an error and/or take other action.

lyakh commented:
@LaurentiuM1234 I agree, that the PR looks rather good now, but I'm going to ask you about one more change, sorry. We discussed again with @lgirdwood and this is how I interpret the result of our discussion, I'd like to be corrected if I misrepresent anything.
First of all - what we understand under "preemptive" pipeline tasks. The example that @lgirdwood used was
```
PCM0.pipeline_1 -> MP3 decoder -\
				 mixer -> DAI
PCM1.pipeline_2 -> volume-------/
```
where pipeline 2 is a normal non-preemptive pipeline, scheduled every 1ms, whereas pipeline 1 takes longer to process its mp3 blocks so it runs every 5ms. But while it's running it can be "preempted" - the execution will switch over to pipeline 2 and when it's done it will yield and execution will switch back to the mp3 decoder.
I don't think such preemptive tasks can be reasonably implemented within the SOF LL scheduler. I think the way to implement them would be by using native Zephyr threads - one Zephyr thread per preemptive SOF pipeline task. This will make them rather expensive but I don't think we have a better option. Then those Zephyr threads will use priorities between 1 (highest) and `CONFIG_NUM_PREEMPT_PRIORITIES` (lowest).
That means, that with your domain implementation you already cut into preemptive tasks:
```
	thread = k_thread_create(&dt->ll_thread,
		 zephyr_preempt_domain_stack[core],
		 ZEPHYR_PDOMAIN_STACK_SIZE,
		 zephyr_preempt_domain_thread_fn,
		 dt, NULL, NULL,
		 CONFIG_NUM_PREEMPT_PRIORITIES - 1,
		 0, K_FOREVER);
```
Which might be not what you want for normal short-running pipeline tasks.
So we think that this PR should separate the DMA domain implementation and any code, related to the future preemptive scheduler. Most of the code in this PR I think actually implements a DMA domain, which effectively means I propose to rename all DMA-domain specific code back to use "dma" instead of "preempt." Any code that should be re-used between this PR and the future preempt scheduler can be extracted to separate files. However, I also think that if at this moment you cannot clearly identify which code will be re-used, it should be ok to just rename everything back to "dma" and later when we implement preemption we extract and re-use any code as appropriate.

LaurentiuM1234 commented:
@lyakh  Makes sense. Will revert back to `dma`-based naming.

> However, I also think that if at this moment you cannot clearly identify which code will be re-used, it should be ok to just rename everything back to "dma" and later when we implement preemption we extract and re-use any code as appropriate.

I believe the best thing to do right now is to wait until the `preempt domain` is actually introduced. This way, the PR will only introduce the DMA domain (as it's supposed to do).

dbaluta commented:
@lgirdwood @lyakh can you suggest an Intel platform that we can test this on? e.g give me the name of the topology and I will enable DMA domain on it and give it a test with CI.

With this we can check if you have the same problems as on i.MX.

dbaluta commented:
@LaurentiuM1234 please remove the 'draft' label from this PR. I think we can merge it as EXPERIMENTAL feature once we test it on some Intel platforms too.

LaurentiuM1234 commented:
**V6 changes**

1) Changed back to `dma`-based naming.

iuliana-prodan reviewed (approved):
This looks good @LaurentiuM1234! Well done!
